### PR TITLE
Refactor the convergence logic

### PR DESCRIPTION
**Pull request by RaunoArike**
This pull request is part of the refactor to simplify the core struct and related methods. In it, I propose the following changes:

- Create new structs for all different convergence criteria, and dispatch based on the types of these structs when calling the `converged()` method. This removes a ton of kwargs from the main struct and makes the code clearer in general. For the more casual users, nothing changes after this refactor: they can still use the `:converge_when` keyword argument to specify a convergence criterion with the default arguments. The more advanced user will need to be a bit more involved than before to define custom convergence criteria: they will have to define one of the new convergence structs on their own and pass that to the `generate_counterfactual()` method. However, the improved clarity of the code seems to justify the added hassle for those users.
- Remove the `converged()` method that had been created specifically for Growing Spheres. There's no reason Growing Spheres shouldn't fit into the pre-existing convergence framework.
- Remove the `:early_stopping` convergence type, which appears to be functionally the same as the `:max_iter` convergence type.

### CODE REVIEW

File: src/counterfactuals/core_struct.jl [hunk 1]
```diff
@@ -30,7 +30,7 @@ end
 		initialization::Symbol = :add_perturbation,
 		generative_model_params::NamedTuple = (;),
 		min_success_rate::AbstractFloat=0.99,
-        converge_when::Symbol=:decision_threshold,
+        convergence::Union{AbstractConvergence,Symbol}=:decision_threshold,
         invalidation_rate::AbstractFloat=0.5,
```
Comment (RaunoArike) [inline src/counterfactuals/core_struct.jl:34]: Redundancies such as the unneccessary keyword arguments here will be removed as part of another PR.

File: src/generators/gradient_based/loss.jl [hunk 2]
```diff
@@ -39,9 +39,15 @@ function ∇(
...
+
+"""
+    hinge_loss(convergence::AbstractConvergence, ce::AbstractCounterfactualExplanation)
+
+The default hinge loss for any convergence criterion.
+Can be overridden inside the `Convergence` module as part of the definition of specific convergence criteria.
+"""
+function hinge_loss(convergence::AbstractConvergence, ce::AbstractCounterfactualExplanation)
+    return 0
 end
```
Comment (VincentPikand) [inline src/generators/gradient_based/loss.jl:53]: I guess we will address this in #371

**Review by VincentPikand — COMMENTED** [review]
LGTM

File: src/generators/gradient_based/loss.jl (hunk 2 above)
Comment (RaunoArike) [reply src/generators/gradient_based/loss.jl:53]: This was my plan, yep

File: src/generate_counterfactual/core.jl [hunk 3]
```diff
@@ -53,7 +53,9 @@ function generate_counterfactual(
     decision_threshold::AbstractFloat=0.5,
     gradient_tol::AbstractFloat=parameters[:τ],
     min_success_rate::AbstractFloat=parameters[:min_success_rate],
-    converge_when::Symbol=:decision_threshold,
+    convergence::Union{AbstractConvergence,Symbol}=Convergence.DecisionThresholdConvergence(;
+        decision_threshold=(1 / length(data.y_levels))
+    ),
```
Comment (kmariuszk) [inline src/generate_counterfactual/core.jl:58]: Doesn't it always overwrites the `convergence` value?

**Review by kmariuszk — CHANGES_REQUESTED** [review]
Well done! Left just one comment that needs to be addressed before I approve it.

File: src/generate_counterfactual/core.jl (hunk 3 above)
Comment (RaunoArike) [reply src/generate_counterfactual/core.jl:58]: I don't think so - it's the same kind of default argument as the ones other keyword arguments here are using. If the user provides a different `convergence` value, that will be used instead. The tests should also fail if this is always used as the `convergence` value. Am I missing something here?

Comment (pat-alt) [reply src/generate_counterfactual/core.jl:58]: This looks fine to me

File: src/generators/gradient_based/loss.jl (hunk 2 above)
Comment (pat-alt) [reply src/generators/gradient_based/loss.jl:53]: Was a about to comment this as well :)

**Review by pat-alt — APPROVED** [review]
Great, this looks good to me

**Review by kmariuszk — APPROVED** [review]
